AG-12606 Screenreader fixes

olegat opened the pull request:
https://ag-grid.atlassian.net/browse/AG-12606

olegat committed:
AG-12606 Fix normaliseMissingAriaAttrs for "ariaLabel" and "ariaChecked"

olegat committed:
AG-12606 Always set initialFocus

The :focus-visible CSS pseudo-class will take care of showing/hidden the
indicator box.

olegat committed:
AG-12606 Add altText to close icon

olegat commented on packages/ag-charts-community/src/dom/elements.ts:
```diff
@@ -8,13 +8,23 @@ interface MissingAriaAttrs {
+// menu item without an accessible text alternative.
+type LabelAndIcon = { label: string; icon?: AgIconName; altText?: undefined };
+type IconOnly = { label?: undefined; icon: AgIconName; altText: string };
+export type LabelIcon = LabelAndIcon | IconOnly;
```
Note: This `LabelIcon` allows 3 syntax modes:

```ts
{ label }         // text only
{ label, icon }   // text & icon
{ icon, altText } // icon only (with altText for screenreaders)
```

Specifying an `icon` without `altText`, or specifying both a `label` and `altText` will raise a compiler error.

olegat commented on packages/ag-charts-community/src/dom/elements.ts:
```diff
@@ -87,6 +97,6 @@ function applyAttrs(element: HTMLElement, attrs?: Attrs) {
-    if (key === 'ariaControls') return 'aria-controls';
-    return key;
+    // Convert camelCaseString to kebab-case-string
+    return key.replace(/([a-z])([A-Z])/g, '$1-$2').toLowerCase();
```
This won't work for some property-attribute pairs like `ariaSetSize` / `aria-setsize` or `ariaPosInSet` / `aria-posinset`. We might want to have a more strict interface for the `Attrs` type.

Thoughts?

lsjroberts replied:
Agreed, the current `Attrs` type also allows function properties like `setEventListener` which is a bit leaky.

If we only care about targetting aria attributes, could we just replace the `aria` with `aria-` and not kebab the whole thing? Or are there other attrs we need to consider?

The reason I called it "missing" aria attrs is the typescript types for elements includes many attrs like `ariaLabel` but not `ariaControls`. Not sure if that is just our ts version being out of date or if it implies something else?

olegat replied:
AG-Grid uses wrappers to set aria attributes: 
https://github.com/ag-grid/ag-grid/blob/419770ca302ce87bc3c9a64b51a464c86132ab7c/community-modules/core/src/utils/aria.ts

We have something similar but it's not so widely used unfortunately:
https://github.com/ag-grid/ag-charts/blob/c2f6bb944c7dd8bce14803cdac6717a485e2537f/packages/ag-charts-community/src/util/attributeUtil.ts

It might be a good idea to start using `attributeUtil.ts` in more places sure as here.

lsjroberts commented on packages/ag-charts-locale/src/en-US.ts:
```diff
@@ -67,6 +67,8 @@ export const AG_CHARTS_LOCALE_EN_US: Record<string, string> = {
     // Alt-text for the 'position-right' icon
     iconAltTextAlignRight: 'Right',
+    // Alt-text for the 'close' icon
+    iconAltTextClose: 'Close',
```
Worth highlighting this with @JamesSwinton - https://github.com/ag-grid/ag-charts/pull/2478

olegat replied:
Noted. Thanks.

alantreadway replied:
We'll hold off merging #2478 until these new keys are merged.

olegat committed:
AG-12606 Add aria-label to tablist

olegat committed:
AG-12606 Add stricter typing to elements.ts functions

The camelCase to kebab-case conversion elements.ts will not correctly work
for property/attribute pairs like ariaSetSize/aria-setsize.

Also, the current type mechanism exposes "attributes" like addEventListener
which is a method not an attribute.

Piggy-back on the existing `setAttribute` function to enforce type safety.

alantreadway reviewed (approved):
LGTM

olegat committed:
AG-12606 Fix e2e test with callouts
